src/experimental/ProbabilisticGraphicalModels/bayesnet.jl

```diff
@@ -166,11 +166,26 @@ Ancestral sampling works by:
 2. Sampling from each node in order, using the already-sampled parent values for conditional distributions
 """
 function ancestral_sampling(bn::BayesianNetwork{V}) where {V}
-    ordered_vertices = Graphs.topological_sort(bn.graph)
-
+    ordered_vertices = Graphs.topological_sort_by_dfs(bn.graph)
     samples = Dict{V,Any}()
 
-    # TODO: Implement sampling logic
+    for vertex_id in ordered_vertices
+        vertex_name = bn.names[vertex_id]
+        if bn.is_observed[vertex_id]
+            samples[vertex_name] = bn.values[vertex_name]
+            continue
+        end
+        if bn.is_stochastic[vertex_id]
+            dist_idx = findfirst(id -> id == vertex_id, bn.stochastic_ids)
```

Comment:
If we have `bn.distributions` to have length `nv`, we can eliminate the needs for `findfirst` which might be bottleneck since every pass is a worst case O(N) _I assume thats how Julia works too_. but also increasing memory usage to O(nv) which might not be too much of a problem, compared to the time complexity saved with removing `findfirst`. One thing that I am not sure is how likely that debugging becomes harder if `undef ` are accessed.